controller/jobs.go

```diff
@@ -231,6 +422,84 @@ func (c *controllerAPI) KillJob(ctx context.Context, w http.ResponseWriter, req
 	}
 }
 
+func (c *controllerAPI) AttachJob(ctx context.Context, w http.ResponseWriter, r *http.Request) {
+	client, job, err := c.connectHost(ctx)
+	if err != nil {
+		respondWithError(w, err)
+		return
+	}
+	var req ct.JobRequest
+	if err := httphelper.DecodeJSON(r, &req); err != nil {
+		respondWithError(w, err)
+		return
+	}
+	attachReq := &host.AttachReq{
+		JobID:  job.ID,
+		Flags:  host.AttachFlagStdout | host.AttachFlagStderr | host.AttachFlagStdin | host.AttachFlagStream,
+		Height: uint16(req.Config.Lines),
+		Width:  uint16(req.Config.Columns),
+	}
+	attachClient, err := client.Attach(attachReq, true)
+	if err != nil {
+		respondWithError(w, fmt.Errorf("attach failed: %s", err.Error()))
+		return
+	}
+	defer attachClient.Close()
+	if err := attachClient.Wait(); err != nil {
+		respondWithError(w, fmt.Errorf("attach wait failed: %s", err.Error()))
+		return
+	}
+	w.Header().Set("Connection", "upgrade")
+	w.Header().Set("Upgrade", "flynn-attach/0")
+	w.WriteHeader(http.StatusSwitchingProtocols)
+	conn, _, err := w.(http.Hijacker).Hijack()
+	if err != nil {
+		panic(err)
+	}
+	defer conn.Close()
+
+	done := make(chan struct{}, 2)
+	cp := func(to io.Writer, from io.Reader) {
+		io.Copy(to, from)
+		done <- struct{}{}
+	}
+	go cp(conn, attachClient.Conn())
+	go cp(attachClient.Conn(), conn)
+	<-done
+	<-done
+}
+
+func (c *controllerAPI) AddJobRequest(ctx context.Context, w http.ResponseWriter, r *http.Request) {
+	var req ct.JobRequest
+	if err := httphelper.DecodeJSON(r, &req); err != nil {
+		respondWithError(w, err)
+		return
+	}
+	req.AppID = c.getApp(ctx).ID
+
+	// check the release exists and set the artifact if necessary
+	releaseData, err := c.releaseRepo.Get(req.ReleaseID)
+	if err != nil {
+		respondWithError(w, err)
+		return
+	}
+	release := releaseData.(*ct.Release)
+	if len(req.ArtifactIDs) == 0 {
+		req.ArtifactIDs = release.ArtifactIDs
+	}
+
+	// TODO: validate
```

Comment:
`JobRequest` needs a JSON schema (you'll likely have to update [controller/schema/schema.go schemaForType](https://github.com/flynn/flynn/blob/3c89a606c28559a5cfa0eadd989c17572e52f03e/controller/schema/schema.go#L55) for `schema.Validate(req)` to work).